### DISCUSSION TITLE
[expo-av] Share bitstream to react-native

### DISCUSSION THREAD

**Discussion by MateWW** [expo/expo#13404] [Proposals and Ideas]
# Summary

Expose audio bitstream to JS, which allows to for example show waveform of the audio.

# Api proposal

| Method | Description |
|---|---|
| bitStreamPrecision | `number // Samples per second`  |
| onPlaybackBitStream | `(bitStream: BitStream) => void` |

```ts
interface BitStream {
  data: Peak[];
  /* Samples per second */
  precision: number;
}

interface Peak {
  /* The geometric mean of all channels  */
  value: number;
}

```

# Motivation

After research, I've noticed there is no valuable library that will allow presenting some audio-related charts like waveforms.

# Drawbacks

Why should we *not* do this?

- With a high level of detail there might exist performance problems.

# Alternatives

We considered creating a facade over expo-av. Sadly, we need to hook on the audio player to inject our code between native and js.

# Adoption strategy

The feature will be optional and shouldn't break anything. 

# How we teach this

We should just adjust documentation as it's an optional feature.

# Unresolved questions

- We need to define what will be the best data granularity to minimize performance issues.
- We need to decide whether we want to share channel data or just geometric mean.

**Comment by brentvatne** [comment]
hey @MateWW - thanks for the proposal! can you elaborate on how the data will be provided? is the idea that it will need to be sent over the bridge? have you considered any other approaches?

**Reply by MateWW** [reply]
Hey @brentvatne, Thanks for the rapid response! In the case of transporting data, We would like to check a load while sending them over the bridge, and then if the result will be insufficient use JSI. 
Sadly I don't have wide experience with JSI so it will probably a little bit more time-consuming than simply use a bridge but I will face the challenge :)

**Reply by brentvatne** [reply]
@MateWW - i suspect going straight to jsi will be the best approach here. we may have someone available soon to lead this implementation, let me get back to you on it next week.

**Reply by MateWW** [reply]
That would be awesome!

**Comment by mrousavy** [comment]
Hi @MateWW!

I'd like to extend this proposal with a few extra details to be more precise.

1. **Data:** We cannot have a single array shared between native and JS using JSI, we still have to have a copy operation. Since Audio is sampled at something like 44100 Hz and each audio frame is represented as 32-bit floats, we have 176.400 bytes (176 kB) of data flowing every second. If an audio file is a minute long, we have a 10.5 MB array in JS - that's going to freeze the JS thread for a while, so accessing all at once (3.) might not be a smart idea. Streaming it (2.) might work without performance issues.
2. **Streaming:** For the `onPlaybackBitStream` function, I'd propose a delivery frequency of 100ms (which is what the AVFoundation framework is outputting anyways), so each 0.1 seconds we get 17 kB of binary audio data (PCM format). This first has to be copied into JSI values, I will benchmark performance on the fly. Currently I would try to simply do this on the normal JS Thread, but we can also talk about doing this in a reanimated "Worklet" (offloaded to a separate Thread), just like how [VisionCamera](https://github.com/mrousavy/react-native-vision-camera) does it in [Frame Processors](https://mrousavy.github.io/react-native-vision-camera/docs/guides/frame-processors).
3. **Accessing it all at once:** For the `Audio.Sound` library we can have a property that can be used to inspect the actual media file. 
    1. `{ channels: { buffer: number[] }[] }` - we could have `Audio.Sound.data` which is a lazy-property returning an object (`AudioData`). The object consists of an array of channels (`channels: Channel[]`), and each `Channel` object contains a `buffer` property.
    2. `number[][]` - If the creation of those nested objects is too slow, I can also simply expose a single array (`Audio.Sound.data`), which is how the AVFoundation API works - it's simply an array of float arrays (`number[][]`).
4. **Recording:** For recording audio we can also expose the same functionality as for audio playback - a callback that gets executed every 0.1 seconds (100ms) with a sample buffer of audio data. Sample size is different per device, but we can cap it to some constant limit to avoid performance issues. The `onRecordedBitStream` function can be used to display a waveform of audio "loudness", just like in the telegram app:

<div align="center">
<img src="https://user-images.githubusercontent.com/15199031/124475818-71494600-dda2-11eb-8f76-8c174c48588d.gif" align="center">
</div>

I've found a great research paper for Audio Visualization using AVFoundation. Currently I can't really estimate how long this project is going to take, since I have to rewrite a few parts of the `EXAVPlayerData` classes. I have to migrate from an `AVQueuePlayer` to a full blown `AVAudioEngine` + `AVAudioPlayerNode` setup to install a tap on the audio bus. Also, in the research paper all visualization operations are run in Metal using Shaders, and they extensively use the Accelerate Framework to improve performance of array operations. (e.g. Calculation of "loudness" of the audio frames) I'm curious to see how slow it would be to run all of this on the CPU instead, and not in Objective-C but rather in JS. Worst case scenario, we can only report audio samples in a much smaller sample size, loosing a lot of precision.

Alternatively, we can abstract a lot of data away from the user. Do we really want to expose the raw `float32[]` of binary audio data? Does the user _need_ this data? Or do we just want to transform the data to an array of "loudness" values ranging from `0.0` to `1.0`? (With samples each e.g. 100 milliseconds) Using this data, the user can then animate changes using REA v2:

```ts
// callback gets called each e.g. 100ms
sound.onPlayback = (stream) => {
  // stream.loudness ranges from 0 to 1
  console.log(`Audio loudness @ ${stream.progress} seconds: ${stream.loudness}`)
  waveformCircleScale.value = withSpring(stream.loudness)
}
sound.play()
```

This of course abstracts a LOT of data away from the user, but at first glance seems to be more performant and simpler.


cc @EvanBacon @ccheever

**Reply by mrousavy** [reply]
Here's the research paper: [**"Audio Visualization in Swift Using Metal and Accelerate (Part 1)"** by Alex Barbulescu](https://betterprogramming.pub/audio-visualization-in-swift-using-metal-accelerate-part-1-390965c095d7)

**Reply by mrousavy** [reply]
With the following code:

```objc
    _engine = [[AVAudioEngine alloc] init];
    [_engine mainMixerNode]; // lazy getter init
    [_engine prepare];
    
    NSError *engineStartError;
    [_engine startAndReturnError:&engineStartError];
    if (engineStartError != nil) {
      NSLog(@"Error starting session!", engineStartError.description);
    } else {
      NSURL *filePath = [[NSBundle mainBundle] URLForResource:@"notification" withExtension:@"wav"];
      _playerNode = [[AVAudioPlayerNode alloc] init];
      
      NSError *fileReadError;
      _audioFile = [[AVAudioFile alloc] initForReading:filePath error:&fileReadError];
      if (fileReadError != nil) {
        NSLog(@"Error reading audio file!", fileReadError.description);
      } else {
        [_engine attachNode:_playerNode];
        [_engine connect:_playerNode to:[_engine mainMixerNode] format:_audioFile.processingFormat];
        
        [_playerNode scheduleFile:_audioFile atTime:nil completionHandler:nil];
        [_playerNode installTapOnBus:0 bufferSize:1024 format:nil block:^(AVAudioPCMBuffer * _Nonnull buffer, AVAudioTime * _Nonnull when) {
          NSLog(@"Frame Length: %u | Capacity: %u | Stride: %u | Time: %u", buffer.frameLength, buffer.frameCapacity, buffer.stride, (UInt32)when.hostTime);
          float* firstChannel = buffer.floatChannelData[0];
          NSLog(@"Frame Buffer received! %f", firstChannel[0]);
        }];
        [_playerNode play];
      }
    }
```

I'm getting the entire buffer from every channel (2 channels in my case) of the audio file every 100ms:

```
2021-07-05 17:34:17.290735+0200 Bare Expo[14844:425071] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2074053928
2021-07-05 17:34:17.290894+0200 Bare Expo[14844:425071] Frame Buffer received! 0.000000
2021-07-05 17:34:17.311940+0200 Bare Expo[14844:425071] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2095387261
2021-07-05 17:34:17.312082+0200 Bare Expo[14844:425071] Frame Buffer received! 0.000000
2021-07-05 17:34:17.386579+0200 Bare Expo[14844:425071] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2174067533
2021-07-05 17:34:17.386709+0200 Bare Expo[14844:425071] Frame Buffer received! 0.000000
2021-07-05 17:34:17.408117+0200 Bare Expo[14844:425071] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2195400866
2021-07-05 17:34:17.408370+0200 Bare Expo[14844:425071] Frame Buffer received! 0.000000
2021-07-05 17:34:17.493276+0200 Bare Expo[14844:425071] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2274051379
2021-07-05 17:34:17.493446+0200 Bare Expo[14844:425071] Frame Buffer received! -0.005432
2021-07-05 17:34:17.514612+0200 Bare Expo[14844:425071] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2295384107
2021-07-05 17:34:17.514743+0200 Bare Expo[14844:425071] Frame Buffer received! -0.005432
2021-07-05 17:34:17.589276+0200 Bare Expo[14844:425071] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2374039471
2021-07-05 17:34:17.589453+0200 Bare Expo[14844:425071] Frame Buffer received! 0.001038
2021-07-05 17:34:17.610687+0200 Bare Expo[14844:425071] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2395372199
2021-07-05 17:34:17.610918+0200 Bare Expo[14844:425071] Frame Buffer received! 0.001038
2021-07-05 17:34:17.685278+0200 Bare Expo[14844:424534] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2474027565
2021-07-05 17:34:17.685448+0200 Bare Expo[14844:424534] Frame Buffer received! -0.098206
2021-07-05 17:34:17.706570+0200 Bare Expo[14844:424534] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2495364050
2021-07-05 17:34:17.706691+0200 Bare Expo[14844:424534] Frame Buffer received! -0.098206
2021-07-05 17:34:17.791864+0200 Bare Expo[14844:424534] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2574028412
2021-07-05 17:34:17.791990+0200 Bare Expo[14844:424534] Frame Buffer received! -0.048523
2021-07-05 17:34:17.813244+0200 Bare Expo[14844:424534] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2595361269
2021-07-05 17:34:17.813357+0200 Bare Expo[14844:424534] Frame Buffer received! -0.048523
2021-07-05 17:34:17.887847+0200 Bare Expo[14844:424534] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2674039782
2021-07-05 17:34:17.887973+0200 Bare Expo[14844:424534] Frame Buffer received! -0.036407
2021-07-05 17:34:17.909255+0200 Bare Expo[14844:424534] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2695372638
2021-07-05 17:34:17.909365+0200 Bare Expo[14844:424534] Frame Buffer received! -0.036407
2021-07-05 17:34:17.984165+0200 Bare Expo[14844:424534] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2774034640
2021-07-05 17:34:17.984483+0200 Bare Expo[14844:424534] Frame Buffer received! 0.040436
2021-07-05 17:34:18.005328+0200 Bare Expo[14844:424534] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2795367677
2021-07-05 17:34:18.005494+0200 Bare Expo[14844:424534] Frame Buffer received! 0.040436
2021-07-05 17:34:18.090554+0200 Bare Expo[14844:425071] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2874033251
2021-07-05 17:34:18.090699+0200 Bare Expo[14844:425071] Frame Buffer received! 0.218781
2021-07-05 17:34:18.111901+0200 Bare Expo[14844:425071] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2895366288
2021-07-05 17:34:18.112021+0200 Bare Expo[14844:425071] Frame Buffer received! 0.218781
2021-07-05 17:34:18.186554+0200 Bare Expo[14844:425071] Frame Length: 4410 | Capacity: 4410 | Stride: 1 | Time: 2974045467
```

I'll try to expose this to JS using JSI and run a few operations on the array to see how slow it is going to get, but the article, which uses Accelerate Framework + Metal and runs all of this on the GPU, already mentions it is _not that fast_:

![Screenshot 2021-07-05 at 18 09 26](https://user-images.githubusercontent.com/15199031/124498125-29371d00-ddbc-11eb-8123-b63e37e9ee41.png)


If it's _too slow_ I'll go with Plan B, so simply exposing an array of numbers which represent the "loudness"/volume of the audio file, one number every ~5ms or so (depending on performance). That's also cool since the users can create waveforms more easily, no expensive array calculations to figure out the average power on all channels, just simple animations as mentioned in my original comment's code sample.

**Reply by MateWW** [reply]
Great work! The only thing I believe is worth mentioning is that In my opinion, we have 2 radically different use-cases. 

1. First is waveform used for example as a progress bar. In this scenario, we need low-quality data up to max screen resolution of items. The problem here is that it has to load the whole file and do some math on it, resulting in possible perf issue for big files. 
 
![image](https://user-images.githubusercontent.com/16048381/124503193-51774980-ddc5-11eb-9ca9-e75898ba9d6d.png)

2. Second is exactly what you've done and for this purpose, we can have much better quality.

![image](https://user-images.githubusercontent.com/16048381/124503671-538dd800-ddc6-11eb-8d20-88683addd0ca.png)

**Reply by MateWW** [reply]
In fact, I've prepared a piece of code for creating waveform like in the first case. It works well except for the time it needs to load over the bridge where it's ~5s for 30s track with 50 peaks per sample. I believe that moving it to JSI will result in much better performance.

**Reply by MateWW** [reply]
So what I can propose is to prepare 2 separate APIs for both scenarios.
Does that make sense?

**Reply by mrousavy** [reply]
Would you mind sharing the code you prepared with the bridge?

**Reply by MateWW** [reply]
Sure here it's. 
It's just loading assets besides the player. Sadly `AVAssetReader` works only for local files :/


```objective-c
- (nullable)_getWaveformData:(AVURLAsset*)avAsset
{
    CMTime currentFrame = avAsset.duration;
    NSError *error;
    AVAssetReader *reader = [AVAssetReader assetReaderWithAsset:avAsset error:&error];
    if(error != nil){
        return nil;
    }
    
    AVAsset *localAsset = reader.asset;
    AVAssetTrack *audioTrack = [[localAsset tracksWithMediaType:AVMediaTypeAudio] objectAtIndex:0];
    NSDictionary *audioSetting = @{
        AVFormatIDKey : [NSNumber numberWithUnsignedInt:kAudioFormatLinearPCM],
        AVLinearPCMBitDepthKey : [NSNumber numberWithInt:16],
        AVLinearPCMIsBigEndianKey: [NSNumber numberWithBool:NO],
        AVLinearPCMIsFloatKey : [NSNumber numberWithBool:NO],
        AVLinearPCMIsNonInterleaved: [NSNumber numberWithBool:NO]
    };
    AVAssetReaderTrackOutput *readerAudioTrackOutput = [AVAssetReaderTrackOutput assetReaderTrackOutputWithTrack:audioTrack outputSettings:audioSetting];
    if ([reader canAddOutput:readerAudioTrackOutput]) {
        [reader addOutput:readerAudioTrackOutput];
    }
    
    UInt32 sampleRate,channelCount;
    NSArray* formatDesc = audioTrack.formatDescriptions;
    for(unsigned int i = 0; i < [formatDesc count]; ++i) {
        CMAudioFormatDescriptionRef item = (__bridge CMAudioFormatDescriptionRef)[formatDesc objectAtIndex:i];
        const AudioStreamBasicDescription* fmtDesc = CMAudioFormatDescriptionGetStreamBasicDescription (item);
        if(fmtDesc ) {
            sampleRate = fmtDesc->mSampleRate;
            channelCount = fmtDesc->mChannelsPerFrame;
        }
    }
    
    UInt32 bytesPerSample = 2 * channelCount;
    SInt16 normalizeMax = 0;
    
    NSMutableArray * fullSongData = [[NSMutableArray alloc] init];
    
    [reader startReading];
    
    UInt64 totalBytes = 0;
    NSMutableArray * totalChnnels = [[NSMutableArray alloc] initWithCapacity:channelCount];
    NSMutableArray * pixelData = [[NSMutableArray alloc] initWithCapacity:channelCount];
    
    NSInteger sampleTally = 0;
    
    NSInteger samplesPerPixel = MAX(64, 0);
    
    while (reader.status == AVAssetReaderStatusReading)
    {
        // Copy the next sample buffer from the reader output.
        CMSampleBufferRef sampleBuffer = [readerAudioTrackOutput copyNextSampleBuffer];
        if (sampleBuffer)
        {
            CMBlockBufferRef blockBufferRef = CMSampleBufferGetDataBuffer(sampleBuffer);
            size_t length = CMBlockBufferGetDataLength(blockBufferRef);
            totalBytes += length;
            @autoreleasepool {
                NSMutableData * data = [NSMutableData dataWithLength:length];
                CMBlockBufferCopyDataBytes(blockBufferRef, 0, length, data.mutableBytes);
                
                SInt16 * samples = (SInt16 *) data.mutableBytes;
                int sampleCount = length / bytesPerSample;
                
                for (int i = 0; i < sampleCount ; i ++) {
                    
                    sampleTally++;
                    
                    for(int x = 0; x<channelCount; x++) {
                        SInt16 peak = *samples++;
                        if([totalChnnels count] < x+1){
                            [totalChnnels addObject: [NSNumber numberWithInt:0]];
                        }
                        if([pixelData count] < x+1){
                            [pixelData addObject: [NSNumber numberWithInt:0]];
                        }
                        SInt64 totalPeak =  [totalChnnels[x] intValue] + peak;
                        totalChnnels[x] = [NSNumber numberWithInt: totalPeak];
                        
                        if (sampleTally > samplesPerPixel) {
                            SInt16 averagePeak = totalPeak/sampleTally;
                            SInt16 fix = abs(averagePeak);
                            if (fix > normalizeMax) {
                                normalizeMax = fix;
                            }
                            pixelData[x] = [NSNumber numberWithShort:averagePeak];
                        }
                    }
                    
                    if (sampleTally > samplesPerPixel) {
                        [fullSongData addObject: [NSArray arrayWithArray:pixelData]];
                        [totalChnnels removeAllObjects];
                        [pixelData removeAllObjects];
                        sampleTally = 0;
                    }
                    
                    if(fullSongData.count >= 30) {
                        self.bitStreamUpdateCallback([NSArray arrayWithArray:fullSongData]);
                        [fullSongData removeAllObjects];
                    }
                }
                
            }
            
            CFRelease(sampleBuffer);
            sampleBuffer = NULL;
        }
    }
    if(fullSongData.count > 0){
        self.bitStreamUpdateCallback(fullSongData);
        [fullSongData removeAllObjects];
    }
}
```

**Reply by MateWW** [reply]
@mrousavy While saying `I'm getting the entire buffer from every channel (2 channels in my case) of the audio file every 100ms:` you think about a piece that is currently playing?

**Reply by mrousavy** [reply]
Yes, streaming audio in general. So either audio playback or microphone recording.

**Reply by mrousavy** [reply]
Okay so I got something working through JSI, and it looks like it runs pretty fast:

```
[Tue Jul 06 2021 11:22:39.492]  LOG      Received sample data! 1 Channels; 4410 Frames; 0
[Tue Jul 06 2021 11:22:39.583]  LOG      Received sample data! 1 Channels; 4410 Frames; -0.00543212890625
[Tue Jul 06 2021 11:22:39.679]  LOG      Received sample data! 1 Channels; 4410 Frames; 0.00103759765625
[Tue Jul 06 2021 11:22:39.776]  LOG      Received sample data! 1 Channels; 4410 Frames; -0.09820556640625
[Tue Jul 06 2021 11:22:39.882]  LOG      Received sample data! 1 Channels; 4410 Frames; -0.04852294921875
[Tue Jul 06 2021 11:22:39.979]  LOG      Received sample data! 1 Channels; 4410 Frames; -0.036407470703125
[Tue Jul 06 2021 11:22:40.740]  LOG      Received sample data! 1 Channels; 4410 Frames; 0.040435791015625
[Tue Jul 06 2021 11:22:40.193]  LOG      Received sample data! 1 Channels; 4410 Frames; 0.218780517578125
[Tue Jul 06 2021 11:22:40.276]  LOG      Received sample data! 1 Channels; 4410 Frames; 0.180206298828125
[Tue Jul 06 2021 11:22:40.382]  LOG      Received sample data! 1 Channels; 4410 Frames; -0.166778564453125
[Tue Jul 06 2021 11:22:40.481]  LOG      Received sample data! 1 Channels; 4410 Frames; -0.1036376953125
[Tue Jul 06 2021 11:22:40.575]  LOG      Received sample data! 1 Channels; 4410 Frames; 0.04388427734375
[Tue Jul 06 2021 11:22:40.682]  LOG      Received sample data! 1 Channels; 4410 Frames; 0.02947998046875
[Tue Jul 06 2021 11:22:40.778]  LOG      Received sample data! 1 Channels; 4410 Frames; -0.002044677734375
[Tue Jul 06 2021 11:22:40.885]  LOG      Received sample data! 1 Channels; 4410 Frames; 0
```

I am creating the following JS object for each time the callback gets invoked:

```ts
interface Channel {
  frames: number[];
}
interface AudioSample {
  channels: Channel[];
}
```

So in my case it's 1 channel, 4410 frames (each being a 64-bit double, so 35kB of memory per sample - that's just the `notification.wav` file though. I'll try with high quality audio later)

If you take a look at the timestamps, they are ~100ms apart, so we are almost not loosing any time at all. The entire JSI object creation, array copying and JS function calling thing takes 1ms-2ms (measured with `CACurrentMediaTime`) - I'll try to implement some Fourier Transforms ([vDSP - Using Fourier Transforms](https://developer.apple.com/library/archive/documentation/Performance/Conceptual/vDSP_Programming_Guide/UsingFourierTransforms/UsingFourierTransforms.html)) purely in JS to see how fast it would be to calculate "average power" (loudness) of the audio in realtime.

Overall this looks pretty good. The biggest challenge would be to partially rewrite the expo-av (EXAVPlayerData) code, since we cannot use `AVQueuePlayer` anymore. We have to use a full blown `AVAudioEngine` + `AVAudioPlayerNode` setup.

Here's the prototype code for that: https://github.com/expo/expo/commit/74f6152163a3f89e080aa96f42bfa545357a9677

Also since I renamed `EXAVPlayerData` from `.m` -> `.mm` my build is failing if I don't comment out the `setStatus:` func:

```
ld: warning: Could not find or use auto-linked framework 'FBSDKCoreKit'
Undefined symbols for architecture x86_64:
  "EXErrorWithMessage(NSString*)", referenced from:
      ___46-[EXAVPlayerData setStatus:resolver:rejecter:]_block_invoke in libEXAV.a(EXAVPlayerData.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

weird.

**Reply by mrousavy** [reply]
too bad the screencapture didn't record the song, but I can assure you it's a banger

https://user-images.githubusercontent.com/15199031/124589752-a31de380-de5a-11eb-8c71-31f94aba9874.mov

 so running the [vDSP_rmsqv algorithm](https://developer.apple.com/documentation/accelerate/1450655-vdsp_rmsqv) purely in JavaScript is extremely fast. I can smoothly animate the circle to represent the "loudness", JS code for this here: https://github.com/expo/expo/commit/04968f8df9ab02f8468e237f8e56360f69ea86dd

**Reply by mrousavy** [reply]
okay so I have laid out an implementation plan.

## iOS

* I need to migrate from `AVQueuePlayer` to a fully fledged `AVAudioEngine` + `AVAudioPlayerNode` setup since you cannot tap into the sample buffer delegate otherwise. There will be a lot of things working different than they work now, so this can be considered a full rewrite of the `Audio.Sound` module. Also `AVAudioEngine` is an extremely finicky API with vague documentation, rarely-helpful error messaging, and almost no online code examples demonstrating more than the most basic tasks...
* For the playback callback, I will expose a `jsi::Function` (something like "`addSampleBufferCallback`") that receives an ID (used to identify the `Audio.Sound` instance) and a callback function which will be invoked every 100ms with new Audio Samples. (`{ channels: { frames: number[] }[] }`)
* For the record callback, this will be a simple parameter to the `startRecording` function. The underlying JS API will also have to delegate that parameter to a native `jsi::Function` since you cannot pass `jsi::Function`s to native bridge modules. That's a bit hacky.

## Android

* On Android, the PCM samples are handled on the C++ side (Audio HAL), and are not exposed to Java (for performance reasons). Instead of partially rewriting the Audio module in C++ (using JNI), we can however use the [Visualizer](https://developer.android.com/reference/android/media/audiofx/Visualizer) API, which receives lower resolution sample buffers, but that's okay for us. It would probably be too slow otherwise. Also, luckily this works with the existing `MediaPlayer` impementation (I believe).
* To expose the methods used for adding a callback, we will need to make the Audio module a Hybrid Object (Java + C++) using fbjni, then looking up the desired instance using it's ID, and finally converting the Java `ByteArray` instances to `jsi::Array`s.

I have created rough prototypes to measure performance (as you can see in the GIF above) and overall it looks really good. The biggest challenge is definitely the partial rewrite of expo-av.

**Reply by MateWW** [reply]
@mrousavy Great work! It's amazing how fast it's. 
What about that second scenario with creating a waveform?
Is there something I can help you with during that implementation?

**Reply by mrousavy** [reply]
The second scenario is already implemented with that approach. You can derive that from the first scenario because the full audio sample buffer is exposed to JS. I have included an algorithm that calculates the "loudness" per timestamp (useful for creating a waveform)

**Reply by danieloi** [reply]
Right now, we can get loudness info in the ‘status’ object while recording audio by default. 
There’s a ‘metering’ flag for that. 
It landed in expo 41. 
A workaround with generating a waveform for prerecorded audio would be to use 
https://github.com/katspaugh/wavesurfer.js/
in a web view and send the levels info you need to RN to render your waveform component. 
If anything, the implementation in that repo might give proven methods/needs for this PR cc @MateWW @mrousavy @brentvatne

**Comment by kennyhalim** [comment]
@brentvatne Hi Brent, I am wondering if its possible to adjust Rode's mic gain level using expo-av library? Our external rode mic can adjust the gain automatically if connected via lightning cable but we are using TRRS cable to connect which does not do it automatically. Any insights you have would be very helpful. Thank you!

**Comment by roeku** [comment]
Did this proposal actually ever get anything through to the expo-av? I'm trying to get this kind of data from expo-av but cannot find any other reference besides this proposal on what I'm trying to do.

**Reply by amanblog** [reply]
Hey! Did you find anything for this? I'm stuck with this too.

**Reply by roeku** [reply]
I started experimenting with building [custom native code](https://docs.expo.dev/workflow/customizing/) but did not get anything working just yet. Other things took priority over this feature for now I'm "faking" it.

**Reply by mihikadutta** [reply]
Unless your project has new deps limitation, there are many tiny-sized rn libraries that support this simple feature.

**Reply by amanblog** [reply]
@roeku Understandable, I am doing the same lol.

**Reply by amanblog** [reply]
@meherhowji Can you point me to some of those libraries? The thing is I just want the data, because the waveform/visualizer is a custom design.

**Reply by brentvatne** [reply]
this is available with [`setOnAudioSampleReceived(callback)`](https://docs.expo.dev/versions/latest/sdk/audio/#setonaudiosamplereceivedcallback) on expo-av. this is a bit of an older project but it is an example of usage: https://github.com/barthap/expo-mega-demo - see [PlayerScreen.tsx](https://github.com/barthap/expo-mega-demo/blob/80dbd1e51923153e51fe51e33d8ec253408bded9/src/screens/PlayerScreen.tsx)

**Reply by amanblog** [reply]
@brentvatne thankyou so much 🙌! Don't know how did I missed this.

**Reply by santiagofm** [reply]
@brentvatne This API only seems to work when you load audio files from a local Asset(and maybe if their remote and it downloads them first) but it doesn't work if you are loading a stream using an url(i.e. online radio) even though the audio buffers.

cc @alanjhughes which seems to be modernizing expo-audio

EDIT: ~Well, indeed AVAudioMix is not supported for remote assets according to whats at the end of https://developer.apple.com/library/archive/qa/qa1716/_index.html~ We workaround this using an `MTAudioProcessingTap`.
Afaict, the `MTAudioProcessingTap` is created and set to the player item fine  but the `process` callback is never called though.

**Reply by smxsm** [reply]
@santiagofm did you find a workaround for remote audio streams? Is that a problem with iOS or Expo? the hook just returns no samples at all for remote streams, which is a bummer 😢

**Reply by smxsm** [reply]
In the new "expo-audio" implementation, it seems "useAudioSampleListener()" is not returning any data yet, even for local files ...?

**Reply by santiagofm** [reply]
@smxsm I did not 😕. I was experimenting with different stuff  and diff libraries but abandoned my efforts hoping the new expo-audio API would address this; it's a bummer to hear it doesn't.

Lmk if you get it working though! I'll prob play around with it next week.

**Reply by smxsm** [reply]
@santiagofm ah, too bad ... I'm afraid I have to give up for now, wasted too much time and Apple makes it really hard to get any audio data from e.g. an AVPlayer object - the only way seems to be a MTAudioProcessingTap like you've mentioned, but I'm not sure there either ... let me know if you get anywhere with this!

**Comment by jacekplocharczyk** [comment]
Hey, is there any update on this in new `expo-audio`?
I'm trying to create visualization during recording audio and stucked :(

**Reply by UchennaOkafor** [reply]
I think what you're looking for is under `metering` and `isMeteringEnabled`. That should give you a value that represents sound activity from the microphone

**Reply by jacekplocharczyk** [reply]
I will take a look :)
Thanks!
